feat: implemented pokemon details

zrnkrystnlee opened the pull request:
- created new connector and vm for the details
- created general details, abilities, moves, stats widgets
- updated the naming conventions of the different models
- updated constants file; separated number, color, and strings constants
- updated overview page - added the search bar
- updated the pokemon actions - added a new action to clear the pokemon details
- updated pubspec.yaml- importing font

App Working:
https://user-images.githubusercontent.com/72370369/228493579-e55d2176-b694-4e88-8c5a-6b9d5c7191da.mp4

Flutter Analyze and Dart Format Result
![image](https://user-images.githubusercontent.com/72370369/228443732-2995a169-85c8-42bc-96f1-e8940c6fbfa1.png)

Updated Analyze and Dart Format Result (3rd Commit)
![image](https://user-images.githubusercontent.com/72370369/228581432-d05638fa-1c99-4bf4-af6e-e3252171d170.png)

zrnkrystnlee committed:
feat: implemented pokemon details
- created new connector and vm for the details
- created general details, abilities, moves, stats widgets
- updated the naming conventions of the different models
- updated constants file; separated number, color, and strings constants
- updated overview page - added the search bar
- updated the pokemon actions - added a new action to clear the pokemon details
- updated pubspec.yaml- importing font

zrnkrystnlee committed:
feat: implemented pokemon details
-(pr feedback)
- Deleted the Search Bar temporary to be implemented in different branch

shawn-ffufm commented on pokedex_async_redux/lib/api/model/moves_details.dart:
```diff
@@ -0,0 +1,15 @@
+@freezed
+class MovesDetails with _$MovesDetails {
+  factory MovesDetails({
+    @JsonKey(name: 'name') required String name, //name of the move
```
do u still need those comments? seems unnecessary already 🙂

shawn-ffufm commented on pokedex_async_redux/lib/api/model/moves_details.dart:
```diff
@@ -0,0 +1,15 @@
+class MovesDetails with _$MovesDetails {
+  factory MovesDetails({
+    @JsonKey(name: 'name') required String name, //name of the move
+    @JsonKey(name: 'url') required String url, //info of the move
```
same here

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/pokemon_details_connector.dart:
```diff
@@ -0,0 +1,27 @@
+import 'package:pokedex_asyn_redux/state/app_state.dart';
+
+class PokemonDetailsConnector extends StatelessWidget {
+  final String pokemonName;
```
seems these properties: pokemonName & pokemonImageUrl, aren't needed anymore since you can already utilize what's inside the `pokemon` object 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/pokemon_details_connector.dart:
```diff
@@ -0,0 +1,27 @@
+      vm: () => PokemonDetailsVmFactory(),
+      onInit: (store) => store.dispatchAsync(GetPokemonDetailsAction(pokemonName: pokemonName)),
+      onDispose: (store) => store.dispatchSync(ClearPokemonDetailsAction()),
+      builder: (context, vm) => PokemonDetailsPage(pokemonDetails: vm.pokemonDetails, pokemon: pokemon),
```
kindly apply extra trailing comma here, also kindly move `pokemon` above `pokemonDetails` 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,113 @@
+import 'package:pokedex_asyn_redux/utils/string_extensions.dart';
+
+class PokemonDetailsPage extends StatelessWidget {
+  const PokemonDetailsPage({Key? key, required this.pokemonDetails, required this.pokemon}) : super(key: key);
```
kindly apply extra trailing comma here, also kindly move `pokemon` above `pokemonDetails` 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,113 @@
+    return Scaffold(
+      backgroundColor: bgDetailsPage,
+      body: pokemonDetails.when(
+        (data) => Stack(
```
kindly rename `data` to `pokemonDetails` instead 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,113 @@
+        ),
+        error: (errorMessage) {
+          WidgetsBinding.instance.addPostFrameCallback(
+            (_) => _showErrorMessageSnackbar(context, errorMessage),
```
if this won't overlaps with the ide's vertical line then no need for this extra trailing comma else keep it as is 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,113 @@
+            (_) => _showErrorMessageSnackbar(context, errorMessage),
+          );
+          return Center(
+            child: Column(
```
do you really still need to use Column widget here or just the Text widget is already enough? 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/abilities_card.dart:
```diff
@@ -0,0 +1,61 @@
+class AbilitiesContainer extends StatelessWidget {
+  const AbilitiesContainer({
+    Key? key,
+    required this.ability,
```
kindly make it a habit to place all required properties at the top 😉

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/general_details_container.dart:
```diff
@@ -0,0 +1,79 @@
+import 'package:pokedex_asyn_redux/widgets/spacing.dart';
+
+class DetailContainer extends StatelessWidget {
+  const DetailContainer({required this.label, required this.value, Key? key}) : super(key: key);
```
kindly observe proper trailing comma here 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/general_details_container.dart:
```diff
@@ -0,0 +1,79 @@
+import 'package:pokedex_asyn_redux/utils/string_constants.dart';
+import 'package:pokedex_asyn_redux/widgets/spacing.dart';
+
+class DetailContainer extends StatelessWidget {
```
kindly put this in a separate dart file 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/general_details_container.dart:
```diff
@@ -0,0 +1,79 @@
+          Text(
+            label,
+            style: const TextStyle(
+              fontWeight: FontWeight.bold,
```
no need for extra trailing comma here 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/general_details_container.dart:
```diff
@@ -0,0 +1,79 @@
+          Text(
+            value,
+            style: const TextStyle(
+              fontWeight: FontWeight.w500,
```
same here

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/moves_card.dart:
```diff
@@ -0,0 +1,69 @@
+import 'package:pokedex_asyn_redux/widgets/spacing.dart';
+
+class MovesContainer extends StatelessWidget {
+  const MovesContainer({Key? key, required this.move}) : super(key: key);
```
kindly add extra trailing comma here 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/moves_card.dart:
```diff
@@ -0,0 +1,69 @@
+      children: [
+        const Text(
+          movesLabel,
+          style: TextStyle(fontSize: 20.0, fontWeight: FontWeight.bold),
```
same here 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/moves_card.dart:
```diff
@@ -0,0 +1,69 @@
+        const VerticalSpace(height: 8.0),
+        SizedBox(
+          height: 240.0,
+          child: SingleChildScrollView(
```
use List view here since pokemon moves are dynamic based on what specific pokemon is opened 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/moves_card.dart:
```diff
@@ -0,0 +1,69 @@
+  }
+}
+
+class MoveItem extends StatelessWidget {
```
move this in a separate dart file as well 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/moves_card.dart:
```diff
@@ -0,0 +1,69 @@
+}
+
+class MoveItem extends StatelessWidget {
+  const MoveItem({Key? key, required this.move}) : super(key: key);
```
kindly add extra trailing comma here 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/stats_card.dart:
```diff
@@ -0,0 +1,120 @@
+class StatsContainer extends StatelessWidget {
+  const StatsContainer({
+    Key? key,
+    required this.stat,
```
see related feedback above about this 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/stats_card.dart:
```diff
@@ -0,0 +1,120 @@
+  Widget build(BuildContext context) {
+    String imageName;
+    String title;
+    switch (stats.stat.name) {
```
would be better if you move this in another separate private function that returns `Map<String, String>` to make this more readable 🙂 

after that then just declare it immediately in a private variable before the `return` statement like this: 

`final something = _getSomething(stats.stat.name);`

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/stats_card.dart:
```diff
@@ -0,0 +1,120 @@
+    String imageName;
+    String title;
+    switch (stats.stat.name) {
+      case 'hp':
```
I bet u know where these hard coded strings should be coming from right? 😝

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_details/widgets/stats_card.dart:
```diff
@@ -0,0 +1,120 @@
+                child: Text(
+                  title,
+                  style: const TextStyle(
+                    fontSize: 12.0,
```
no need for extra trailing comma 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/widgets/pokemon_card.dart:
```diff
@@ -16,80 +18,93 @@ class PokemonCard extends StatelessWidget {
-              textAlign: TextAlign.center,
+    return GestureDetector(
+      onTap: () {
+        Navigator.of(context).push(
```
kindly move this into a private function to make this more readable 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/widgets/pokemon_card.dart:
```diff
@@ -16,80 +18,93 @@ class PokemonCard extends StatelessWidget {
+      child: Card(
+        margin: const EdgeInsets.all(commonSpaceSize),
+        shape: RoundedRectangleBorder(
+          borderRadius: BorderRadius.circular(15.0),
```
no need for extra trailing comma 🙂

shawn-ffufm commented on pokedex_async_redux/lib/utils/color_constants.dart:
```diff
@@ -3,5 +3,9 @@ import 'package:flutter/material.dart';
 /// //Pokemon Overview Page Color Constants
 const Color lightColor = Colors.white;
 const Color darkColor = Colors.black;
+const Color greyColor = Colors.grey;
```
kindly rename this dart file to `colors.dart` 🙂

shawn-ffufm commented on pokedex_async_redux/lib/utils/color_constants.dart:
```diff
@@ -3,5 +3,9 @@ import 'package:flutter/material.dart';
+const Color greyColor = Colors.grey;
 const Color typeColorContainer = Colors.blueGrey;
-const Color loadingScreen = Color.fromRGBO(254, 159, 78, 1);
+const Color mainBgColor = Color.fromRGBO(254, 159, 78, 1);
```
also no need to indicate its Color type, same goes for all in this entire file 🙂

shawn-ffufm commented on pokedex_async_redux/lib/utils/string_constants.dart:
```diff
@@ -0,0 +1,46 @@
+const String typePrimary = 'Type 1';
+
+///Pokemon Loading and Error Message
+const String pikachuRunningImage = 'assets/images/pikachu-running.gif';
```
also here no need to indicate its String type, same goes for all in this entire file 🙂

shawn-ffufm commented on pokedex_async_redux/lib/widgets/reusable_appbar.dart:
```diff
@@ -0,0 +1,27 @@
+import 'package:pokedex_asyn_redux/utils/color_constants.dart';
+
+class CustomAppBar extends StatelessWidget implements PreferredSizeWidget {
+  final String title;
```
move this below the constructor 🙂

shawn-ffufm commented on pokedex_async_redux/lib/widgets/reusable_appbar.dart:
```diff
@@ -0,0 +1,27 @@
+class CustomAppBar extends StatelessWidget implements PreferredSizeWidget {
+  final String title;
+
+  const CustomAppBar({required this.title, Key? key}) : super(key: key);
```
kindly observe extra trailing comma 🙂

shawn-ffufm commented on pokedex_async_redux/lib/widgets/reusable_appbar.dart:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  @override
+  Size get preferredSize => const Size.fromHeight(kToolbarHeight);
```
do you still need this? if not kindly get rid as well on implementing or extending its class 🙂

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

zrnkrystnlee committed:
feat: implemented pokemon details
(pr feedback)
- added hard coded strings to constants
- applied extra trailing comma
- changed SingleChildScrollView to a ListView
- created private function for onTap
- moved required properties before Key
- remove unecessary comments, trailing commas, color/string type
- separated classes which are in the same dart file
- separated to a private function the switch case in stats and return as Map<String, String>
